File: nisos/when.py
"""Spoken dates and times, turned into a real ``datetime``.

«αύριο στις πέντε» and "tomorrow at five" both have to become a specific
moment before anything can be written into a calendar. This module is the only
place that knows how, and it is deliberately pure: text in, a datetime out, no
phone, no clock app, no Android. It runs and tests on a laptop.

Three decisions worth knowing before you change anything here
------------------------------------------------------------
**A bare hour between one and seven means the afternoon.** «στις πέντε» is
17:00, not 05:00, because nobody arranges a dentist for five in the morning
and says it that casually. Eight through twelve stay as spoken. Saying «το
πρωί» / "in the morning" forces AM either way, and «το βράδυ» / "pm" forces
PM. One rule, always the same, easy to say out loud when it gets it wrong.

**A time with no day is the next time that time happens.** Today if it is
still ahead, tomorrow if it has passed. A day with no time is
:data:`DEFAULT_HOUR`, because "put the dentist in for Thursday" is a real
sentence and refusing it would be pedantic.

**Word positions, not character offsets.** :func:`parse` reports which *words*
it consumed, so the caller can take what is left as the title. Character
offsets would be wrong: flattening a transcript is not length-preserving (an
accent that arrives as a separate combining character costs a character and no
words), but it never splits or joins a word.

Extending
---------
A new phrasing is usually one entry in one of the tables below. Everything is
written in **normalised** form -- lowercase, unaccented, final sigma as plain
sigma -- so run a candidate through :func:`nisos.normalise.normalise` before
adding it.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta

from .normalise import NUMBER_WORDS

__all__ = ["When", "parse", "DEFAULT_HOUR", "DEFAULT_MINUTES", "AFTERNOON_UNTIL"]


#: What "Thursday", with no time attached, means.
DEFAULT_HOUR = 9

#: How long an appointment lasts when nobody says.
DEFAULT_MINUTES = 60

#: A bare hour at or below this is read as afternoon: «στις πέντε» is 17:00.
#: Above it, the hour is taken at face value -- «στις δέκα» stays 10:00.
AFTERNOON_UNTIL = 7


# --------------------------------------------------------------------------
# The tables
# --------------------------------------------------------------------------

#: Whole words meaning a day, relative to today.
DAY_WORDS: dict[str, dict[str, int]] = {
    "el": {"σημερα": 0, "αποψε": 0, "αυριο": 1, "μεθαυριο": 2},
    "en": {"today": 0, "tonight": 0, "tomorrow": 1},
}

#: Day words that also carry a time of day. «απόψε στις οκτώ» is 20:00.
EVENING_DAYS = {"αποψε", "tonight"}

#: Weekday **stems**, matched with startswith, so «τη Δευτέρα», «της Δευτέρας»
#: and "on Mondays" all land on the same day.
WEEKDAYS: dict[str, dict[str, int]] = {
    "el": {"δευτερα": 0, "τριτη": 1, "τεταρτη": 2, "πεμπτη": 3,
           "παρασκευη": 4, "σαββατο": 5, "κυριακη": 6},
    "en": {"monday": 0, "tuesday": 1, "wednesday": 2, "thursday": 3,
           "friday": 4, "saturday": 5, "sunday": 6},
}

#: Words that force the morning, and words that force the afternoon or later.
AM_WORDS = {"el": ("πρωι", "πρωινο", "πρωια"), "en": ("am", "morning")}
PM_WORDS = {"el": ("απογευμα", "απογευματινο", "βραδυ", "βραδι", "μεσημερι",
                   "αποψε"),
            "en": ("pm", "afternoon", "evening", "night", "tonight", "noon")}

#: Words that introduce a clock time, so the number after them is an hour and
#: not a duration, a date or a house number.
AT_WORDS = {"el": ("στισ", "στη", "στην", "στο", "στον", "ωρα"),
            "en": ("at",)}

#: Words that introduce a duration.
FOR_WORDS = {"el": ("για",), "en": ("for",)}

#: Duration unit **stems**: «λεπτά», «λεπτό», "minutes", "min".
UNIT_MINUTES = {"el": (("ωρ", 60), ("λεπτ", 1)),
                "en": (("hour", 60), ("hr", 60), ("minute", 1), ("min", 1))}

#: Half and quarter, in the two shapes Greek actually says them.
HALF_WORDS = {"el": ("μιση", "μισι"), "en": ("half",)}
QUARTER_WORDS = {"el": ("τεταρτο",), "en": ("quarter",)}

#: «πέντε παρά τέταρτο» -- a quarter *to* rather than *past*.
TO_WORDS = {"el": ("παρα",), "en": ("to",)}

#: «και μισή» -- the joiner between an hour and its minutes. Note this is the
#: same word the router splits multi-action commands on; that is safe, because
#: a split is only accepted when *every* piece routes on its own, and «μισή»
#: does not route to anything.
AND_WORDS = {"el": ("και", "κι"), "en": ("and",)}

_PUNCTUATION = " ,.;:!?«»\"'()"


def _bare(word: str) -> str:
    """A word with the punctuation a recogniser sprinkles on it removed."""
    return word.strip(_PUNCTUATION)


def _order(language: str) -> list[str]:
    """The language to check first, then the other one.

    Code-switching is normal here -- "meeting με τον Νίκο tomorrow" is a
    sentence people say -- so nothing is ever checked in one language only.
    """
    first = language if language in DAY_WORDS else "en"
    return [first] + [other for other in DAY_WORDS if other != first]


def _lookup(word: str, table: dict[str, dict[str, int]], language: str):
    """Find `word` in a per-language table of whole words."""
    for lang in _order(language):
        if word in table.get(lang, {}):
            return table[lang][word]
    return None


def _weekday(word: str, language: str) -> int | None:
    """Find `word` in the weekday table, matching on the stem."""
    for lang in _order(language):
        for stem, index in WEEKDAYS.get(lang, {}).items():
            if word.startswith(stem):
                return index
    return None


def _in(word: str, table: dict[str, tuple[str, ...]], language: str) -> bool:
    """True if `word` appears in a per-language tuple of words."""
    return any(word in table.get(lang, ()) for lang in _order(language))


def _number(word: str, language: str) -> int | None:
    """Read a word as a number: digits first, then spelled out."""
    if word.isdigit():
        return int(word)
    for lang in _order(language):
        value = NUMBER_WORDS.get(lang, {}).get(word)
        if value is not None:
            return value
    return None


def _clock(word: str, language: str) -> tuple[int, int] | None:
    """Read a word as a clock time.

    Handles ``17:30``, ``17.30`` and ``5``, plus the spelled-out hours, and
    the English ``5pm`` that arrives as one token.

    Returns:
        ``(hour, minute)``, or None if the word is not a time at all.
    """
    suffix = None
    if word.endswith(("am", "pm")) and len(word) > 2:
        suffix, word = word[-2:], word[:-2]

    for separator in (":", ".", "h"):
        if separator in word:
            left, _, right = word.partition(separator)
            if left.isdigit() and right.isdigit():
                hour, minute = int(left), int(right)
                break
    else:
        value = _number(word, language)
        if value is None:
            return None
        hour, minute = value, 0

    if suffix == "pm" and hour < 12:
        hour += 12
    elif suffix == "am" and hour == 12:
        hour = 0

    if 0 <= hour <= 23 and 0 <= minute <= 59:
        return hour, minute
    return None


# --------------------------------------------------------------------------
# The result
# --------------------------------------------------------------------------

@dataclass(frozen=True)
class When:
    """A moment, how long it lasts, and which words paid for it.

    Attributes:
        start: The resolved local datetime.
        minutes: Duration, defaulting to :data:`DEFAULT_MINUTES`.
        words: Indices of the words :func:`parse` consumed. Whatever is left
            over is the caller's -- for ``calendar.add`` it is the title.
    """

    start: datetime
    minutes: int = DEFAULT_MINUTES
    words: frozenset[int] = frozenset()

    def iso(self) -> str:
        """``YYYY-MM-DDTHH:MM`` in local time.

        Deliberately no timezone and no seconds. This string is the contract
        between the router, the model and :func:`nisos.actions.calendar_add`,
        and a model asked for an appointment "tomorrow at five" knows what
        local wall-clock time means and does not know the phone's offset.
        """
        return self.start.strftime("%Y-%m-%dT%H:%M")


# --------------------------------------------------------------------------
# The parse
# --------------------------------------------------------------------------

def parse(words: list[str], language: str = "en",
          now: datetime | None = None) -> When | None:
    """Find a date and time in an already-normalised list of words.

    Args:
        words: The utterance, split on whitespace, already through
            :func:`nisos.normalise.normalise`.
        language: ``"el"`` or ``"en"``. Only decides which table is checked
            first -- both are always checked, because people code-switch.
        now: The moment to resolve relative words against. Injected so the
            tests are not a lottery at midnight.

    Returns:
        A :class:`When`, or None when the words contain no time at all --
        which is a normal answer, not a failure.

    >>> from datetime import datetime
    >>> monday = datetime(2026, 8, 10, 12, 0)
    >>> parse("αυριο στισ πεντε".split(), "el", monday).iso()
    '2026-08-11T17:00'
    >>> parse("tomorrow at 9 am".split(), "en", monday).iso()
    '2026-08-11T09:00'
    >>> parse("dentist".split(), "en", monday) is None
    True
    """
    now = now or datetime.now()

    day_offset: int | None = None
    weekday: int | None = None
    hour: int | None = None
    minute = 0
    meridiem: str | None = None
    minutes: int | None = None
    used: set[int] = set()

    index = 0
    while index < len(words):
        word = _bare(words[index])
        if not word:
            index += 1
            continue

        # -- "for 30 minutes", «για μία ώρα» --------------------------------
        if _in(word, FOR_WORDS, language):
            span = _duration(words, index + 1, language)
            if span:
                minutes, consumed = span
                used.update({index, *consumed})
                index = max(consumed) + 1
                continue

        # -- today / tomorrow / tonight -------------------------------------
        offset = _lookup(word, DAY_WORDS, language)
        if offset is not None:
            day_offset = offset
            if word in EVENING_DAYS and meridiem is None:
                meridiem = "pm"
            used.add(index)
            index += 1
            continue

        # -- Monday, «τη Δευτέρα» -------------------------------------------
        found = _weekday(word, language)
        if found is not None:
            weekday = found
            used.add(index)
            index += 1
            continue

        # -- morning / afternoon / pm ---------------------------------------
        if _in(word, AM_WORDS, language):
            meridiem = "am"
            used.add(index)
            index += 1
            continue
        if _in(word, PM_WORDS, language):
            meridiem = "pm"
            used.add(index)
            index += 1
            continue

        # -- "half past five" (English puts it in front) --------------------
        if (hour is None and _in(word, HALF_WORDS, language)
                and index + 2 < len(words)
                and _bare(words[index + 1]) == "past"):
            clock = _clock(_bare(words[index + 2]), language)
            if clock:
                hour, minute = clock[0], 30
                used.update({index, index + 1, index + 2})
                index += 3
                continue

        # -- «στις πέντε», "at 5", "at 17:30" -------------------------------
        if _in(word, AT_WORDS, language) and index + 1 < len(words):
            clock = _clock(_bare(words[index + 1]), language)
            if clock:
                hour, minute = clock
                used.update({index, index + 1})
                index += 2
                hour, minute = _adjust(words, index, language, hour, minute,
                                       used)
                index = _skip_used(index, used)
                continue

        # -- a bare 17:30, which needs no introduction ----------------------
        if hour is None and (":" in word or "." in word) and _clock(word, language):
            hour, minute = _clock(word, language)  # type: ignore[misc]
            used.add(index)
            index += 1
            hour, minute = _adjust(words, index, language, hour, minute, used)
            index = _skip_used(index, used)
            continue

        index += 1

    if hour is None and day_offset is None and weekday is None:
        return None

    start = _resolve(now, day_offset, weekday, hour, minute, meridiem)
    return When(start=start,
                minutes=minutes if minutes is not None else DEFAULT_MINUTES,
                words=frozenset(used))


def _skip_used(index: int, used: set[int]) -> int:
    """Move past any words a lookahead has already claimed."""
    while index in used:
        index += 1
    return index


def _adjust(words: list[str], index: int, language: str, hour: int,
            minute: int, used: set[int]) -> tuple[int, int]:
    """Apply «και μισή», «και τέταρτο» and «παρά τέταρτο» to an hour.

    Args:
        index: Where to start looking, immediately after the hour.
        used: Updated in place with anything consumed.

    Returns:
        The adjusted ``(hour, minute)``.
    """
    if index >= len(words):
        return hour, minute

    word = _bare(words[index])
    following = _bare(words[index + 1]) if index + 1 < len(words) else ""

    if _in(word, AND_WORDS, language) and following:
        if _in(following, HALF_WORDS, language):
            used.update({index, index + 1})
            return hour, 30
        if _in(following, QUARTER_WORDS, language):
            used.update({index, index + 1})
            return hour, 15

    if _in(word, TO_WORDS, language) and _in(following, QUARTER_WORDS, language):
        used.update({index, index + 1})
        return (hour - 1) % 24, 45

    return hour, minute


def _duration(words: list[str], index: int, language: str
              ) -> tuple[int, set[int]] | None:
    """Read "30 minutes" / «μισή ώρα» / «μία ώρα» starting at `index`.

    Returns:
        ``(minutes, consumed indices)``, or None if what follows is not a
        duration -- "for Anna" must not become an appointment length.
    """
    if index >= len(words):
        return None

    consumed = {index}
    word = _bare(words[index])

    if _in(word, HALF_WORDS, language):
        count: float = 0.5
    else:
        value = _number(word, language)
        if value is None:
            return None
        count = value

    if index + 1 >= len(words):
        return None
    unit = _bare(words[index + 1])
    for lang in _order(language):
        for stem, size in UNIT_MINUTES.get(lang, ()):
            if unit.startswith(stem):
                consumed.add(index + 1)
                return max(1, round(count * size)), consumed
    return None


def _resolve(now: datetime, day_offset: int | None, weekday: int | None,
             hour: int | None, minute: int, meridiem: str | None) -> datetime:
    """Turn the pieces into one moment.

    The rules, in the order they are applied:

    1. A bare hour of one to seven with nothing to say otherwise means the
       afternoon -- see :data:`AFTERNOON_UNTIL`.
    2. No hour at all means :data:`DEFAULT_HOUR`.
    3. A named weekday means its next occurrence, today included only if the
       time has not passed.
    4. No day at all means today, rolled to tomorrow if the time has passed.
    """
    if hour is None:
        hour, minute = DEFAULT_HOUR, 0
    elif meridiem == "pm" and hour < 12:
        hour += 12
    elif meridiem == "am" and hour == 12:
        hour = 0
    elif meridiem is None and 1 <= hour <= AFTERNOON_UNTIL:
        hour += 12

    start = now.replace(hour=hour, minute=minute, second=0, microsecond=0)

    if weekday is not None:
        ahead = (weekday - start.weekday()) % 7
        if ahead == 0 and start <= now:
            ahead = 7
        return start + timedelta(days=ahead)

    if day_offset is not None:
        return start + timedelta(days=day_offset)

    if start <= now:
        return start + timedelta(days=1)
    return start

File: android/app/src/main/java/app/nisos/core/When.kt
package app.nisos.core

import java.time.LocalDateTime

/**
 * Spoken dates and times, turned into a real [LocalDateTime].
 *
 * A port of `nisos/when.py`. «αύριο στις πέντε» and "tomorrow at five" both
 * have to become a specific moment before anything can go in a calendar. Pure
 * by design: text in, a datetime out, no Android anywhere.
 *
 * Three decisions worth knowing before changing anything here
 * ----------------------------------------------------------
 * **A bare hour between one and seven means the afternoon.** «στις πέντε» is
 * 17:00, not 05:00, because nobody arranges a dentist for five in the morning
 * and says it that casually. Eight through twelve stay as spoken. «το πρωί» /
 * "in the morning" forces AM either way, and «το βράδυ» / "pm" forces PM. One
 * rule, always the same, easy to say out loud when it gets it wrong.
 *
 * **A time with no day is the next time that time happens.** Today if it is
 * still ahead, tomorrow if it has passed. A day with no time is [DEFAULT_HOUR],
 * because "put the dentist in for Thursday" is a real sentence.
 *
 * **Word positions, not character offsets.** [parse] reports which *words* it
 * consumed, so the caller can take what is left as the title.
 */

/** What "Thursday", with no time attached, means. */
const val DEFAULT_HOUR = 9

/** How long an appointment lasts when nobody says. */
const val DEFAULT_MINUTES = 60

/** A bare hour at or below this is read as afternoon: «στις πέντε» is 17:00. */
const val AFTERNOON_UNTIL = 7

private val DAY_WORDS = mapOf(
    "el" to mapOf("σημερα" to 0, "αποψε" to 0, "αυριο" to 1, "μεθαυριο" to 2),
    "en" to mapOf("today" to 0, "tonight" to 0, "tomorrow" to 1),
)

/** Day words that also carry a time of day. «απόψε στις οκτώ» is 20:00. */
private val EVENING_DAYS = setOf("αποψε", "tonight")

/** Weekday **stems**, matched with startsWith, so «της Δευτέρας» still lands. */
private val WEEKDAYS = mapOf(
    "el" to mapOf(
        "δευτερα" to 1, "τριτη" to 2, "τεταρτη" to 3, "πεμπτη" to 4,
        "παρασκευη" to 5, "σαββατο" to 6, "κυριακη" to 7,
    ),
    "en" to mapOf(
        "monday" to 1, "tuesday" to 2, "wednesday" to 3, "thursday" to 4,
        "friday" to 5, "saturday" to 6, "sunday" to 7,
    ),
)

private val AM_WORDS = mapOf(
    "el" to setOf("πρωι", "πρωινο", "πρωια"),
    "en" to setOf("am", "morning"),
)
private val PM_WORDS = mapOf(
    "el" to setOf("απογευμα", "απογευματινο", "βραδυ", "βραδι", "μεσημερι", "αποψε"),
    "en" to setOf("pm", "afternoon", "evening", "night", "tonight", "noon"),
)

/** Words that introduce a clock time, so the number after one is an hour. */
private val AT_WORDS = mapOf(
    "el" to setOf("στισ", "στη", "στην", "στο", "στον", "ωρα"),
    "en" to setOf("at"),
)
private val FOR_WORDS = mapOf("el" to setOf("για"), "en" to setOf("for"))

/** Duration unit **stems**: «λεπτά», «λεπτό», "minutes", "min". */
private val UNIT_MINUTES = mapOf(
    "el" to listOf("ωρ" to 60, "λεπτ" to 1),
    "en" to listOf("hour" to 60, "hr" to 60, "minute" to 1, "min" to 1),
)

private val HALF_WORDS = mapOf("el" to setOf("μιση", "μισι"), "en" to setOf("half"))
private val QUARTER_WORDS = mapOf("el" to setOf("τεταρτο"), "en" to setOf("quarter"))
private val TO_WORDS = mapOf("el" to setOf("παρα"), "en" to setOf("to"))

/**
 * «και μισή» -- the joiner between an hour and its minutes.
 *
 * Note this is the same word the router splits multi-action commands on. That
 * is safe: a split is only accepted when *every* piece routes on its own, and
 * «μισή» routes to nothing.
 */
private val AND_WORDS = mapOf("el" to setOf("και", "κι"), "en" to setOf("and"))

private const val PUNCTUATION = " ,.;:!?«»\"'()"

private fun bare(word: String) = word.trim { it in PUNCTUATION }

/** The language to check first, then the other. People code-switch. */
private fun order(language: String): List<String> {
    val first = if (DAY_WORDS.containsKey(language)) language else "en"
    return listOf(first) + DAY_WORDS.keys.filter { it != first }
}

private fun inTable(word: String, table: Map<String, Set<String>>, language: String) =
    order(language).any { table[it]?.contains(word) == true }

private fun weekdayOf(word: String, language: String): Int? {
    for (lang in order(language)) {
        WEEKDAYS[lang]?.forEach { (stem, index) ->
            if (word.startsWith(stem)) return index
        }
    }
    return null
}

private fun numberOf(word: String, language: String): Int? {
    word.toIntOrNull()?.let { return it }
    for (lang in order(language)) {
        NUMBER_WORDS[lang]?.get(word)?.let { return it }
    }
    return null
}

/**
 * Read a word as a clock time: `17:30`, `17.30`, `5`, `5pm`, or «πέντε».
 *
 * Returns hour to minute, or null if the word is not a time at all.
 */
private fun clockOf(raw: String, language: String): Pair<Int, Int>? {
    var word = raw
    var suffix: String? = null
    if (word.length > 2 && (word.endsWith("am") || word.endsWith("pm"))) {
        suffix = word.takeLast(2)
        word = word.dropLast(2)
    }

    var hour: Int
    var minute = 0
    val separator = listOf(":", ".", "h").firstOrNull { word.contains(it) }
    if (separator != null) {
        val (left, right) = word.split(separator, limit = 2).let { it[0] to it.getOrElse(1) { "" } }
        val h = left.toIntOrNull() ?: return null
        val m = right.toIntOrNull() ?: return null
        hour = h
        minute = m
    } else {
        hour = numberOf(word, language) ?: return null
    }

    if (suffix == "pm" && hour < 12) hour += 12
    if (suffix == "am" && hour == 12) hour = 0

    return if (hour in 0..23 && minute in 0..59) hour to minute else null
}

/**
 * A moment, how long it lasts, and which words paid for it.
 *
 * @property words indices of the words [parse] consumed. Whatever is left over
 *   is the caller's -- for `calendar.add` it is the title.
 */
data class When(
    val start: LocalDateTime,
    val minutes: Int = DEFAULT_MINUTES,
    val words: Set<Int> = emptySet(),
) {
    /**
     * `YYYY-MM-DDTHH:MM` in local time.
     *
     * Deliberately no timezone and no seconds. This string is the contract
     * between the router, the model and `calendar.add`, and a model asked for
     * "tomorrow at five" knows what local wall-clock time means and does not
     * know the phone's offset.
     */
    fun iso(): String = String.format(
        java.util.Locale.ROOT, "%04d-%02d-%02dT%02d:%02d",
        start.year, start.monthValue, start.dayOfMonth, start.hour, start.minute,
    )
}

/**
 * Find a date and time in an already-normalised list of words.
 *
 * @param now injected so the tests are not a lottery at midnight.
 * @return null when the words contain no time at all, which is a normal answer
 *   and not a failure.
 */
fun parseWhen(words: List<String>, language: String = "en", now: LocalDateTime): When? {
    var dayOffset: Int? = null
    var weekday: Int? = null
    var hour: Int? = null
    var minute = 0
    var meridiem: String? = null
    var minutes: Int? = null
    val used = mutableSetOf<Int>()

    var index = 0
    while (index < words.size) {
        val word = bare(words[index])
        if (word.isEmpty()) { index++; continue }

        // -- "for 30 minutes", «για μία ώρα» ---------------------------------
        if (inTable(word, FOR_WORDS, language)) {
            val span = duration(words, index + 1, language)
            if (span != null) {
                minutes = span.first
                used.add(index)
                used.addAll(span.second)
                index = span.second.max() + 1
                continue
            }
        }

        // -- today / tomorrow / tonight --------------------------------------
        val offset = order(language).firstNotNullOfOrNull { DAY_WORDS[it]?.get(word) }
        if (offset != null) {
            dayOffset = offset
            if (word in EVENING_DAYS && meridiem == null) meridiem = "pm"
            used.add(index); index++; continue
        }

        // -- Monday, «τη Δευτέρα» --------------------------------------------
        val named = weekdayOf(word, language)
        if (named != null) {
            weekday = named
            used.add(index); index++; continue
        }

        // -- morning / afternoon / pm ----------------------------------------
        if (inTable(word, AM_WORDS, language)) {
            meridiem = "am"; used.add(index); index++; continue
        }
        if (inTable(word, PM_WORDS, language)) {
            meridiem = "pm"; used.add(index); index++; continue
        }

        // -- "half past five" (English puts it in front) ----------------------
        if (hour == null && inTable(word, HALF_WORDS, language) &&
            index + 2 < words.size && bare(words[index + 1]) == "past"
        ) {
            val clock = clockOf(bare(words[index + 2]), language)
            if (clock != null) {
                hour = clock.first
                minute = 30
                used.addAll(listOf(index, index + 1, index + 2))
                index += 3
                continue
            }
        }

        // -- «στις πέντε», "at 5", "at 17:30" ---------------------------------
        if (inTable(word, AT_WORDS, language) && index + 1 < words.size) {
            val clock = clockOf(bare(words[index + 1]), language)
            if (clock != null) {
                hour = clock.first
                minute = clock.second
                used.addAll(listOf(index, index + 1))
                index += 2
                val adjusted = adjust(words, index, language, hour!!, minute, used)
                hour = adjusted.first
                minute = adjusted.second
                while (index in used) index++
                continue
            }
        }

        // -- a bare 17:30, which needs no introduction ------------------------
        if (hour == null && (word.contains(":") || word.contains("."))) {
            val clock = clockOf(word, language)
            if (clock != null) {
                hour = clock.first
                minute = clock.second
                used.add(index)
                index++
                val adjusted = adjust(words, index, language, hour!!, minute, used)
                hour = adjusted.first
                minute = adjusted.second
                while (index in used) index++
                continue
            }
        }

        index++
    }

    if (hour == null && dayOffset == null && weekday == null) return null

    return When(
        start = resolve(now, dayOffset, weekday, hour, minute, meridiem),
        minutes = minutes ?: DEFAULT_MINUTES,
        words = used,
    )
}

/** Apply «και μισή», «και τέταρτο» and «παρά τέταρτο» to an hour. */
private fun adjust(
    words: List<String>, index: Int, language: String,
    hour: Int, minute: Int, used: MutableSet<Int>,
): Pair<Int, Int> {
    if (index >= words.size) return hour to minute

    val word = bare(words[index])
    val following = if (index + 1 < words.size) bare(words[index + 1]) else ""

    if (inTable(word, AND_WORDS, language) && following.isNotEmpty()) {
        if (inTable(following, HALF_WORDS, language)) {
            used.addAll(listOf(index, index + 1)); return hour to 30
        }
        if (inTable(following, QUARTER_WORDS, language)) {
            used.addAll(listOf(index, index + 1)); return hour to 15
        }
    }
    if (inTable(word, TO_WORDS, language) && inTable(following, QUARTER_WORDS, language)) {
        used.addAll(listOf(index, index + 1))
        return ((hour - 1 + 24) % 24) to 45
    }
    return hour to minute
}

/**
 * Read "30 minutes" / «μισή ώρα» / «μία ώρα» starting at [index].
 *
 * Returns null if what follows is not a duration -- "for Anna" must not become
 * an appointment length.
 */
private fun duration(words: List<String>, index: Int, language: String): Pair<Int, Set<Int>>? {
    if (index >= words.size) return null
    val consumed = mutableSetOf(index)
    val word = bare(words[index])

    val count: Double = if (inTable(word, HALF_WORDS, language)) {
        0.5
    } else {
        (numberOf(word, language) ?: return null).toDouble()
    }

    if (index + 1 >= words.size) return null
    val unit = bare(words[index + 1])
    for (lang in order(language)) {
        UNIT_MINUTES[lang]?.forEach { (stem, size) ->
            if (unit.startsWith(stem)) {
                consumed.add(index + 1)
                return maxOf(1, Math.round(count * size).toInt()) to consumed
            }
        }
    }
    return null
}

/**
 * Turn the pieces into one moment.
 *
 * 1. A bare hour of one to seven means the afternoon -- see [AFTERNOON_UNTIL].
 * 2. No hour at all means [DEFAULT_HOUR].
 * 3. A named weekday means its next occurrence, today included only if the
 *    time has not passed.
 * 4. No day at all means today, rolled to tomorrow if the time has passed.
 */
private fun resolve(
    now: LocalDateTime, dayOffset: Int?, weekday: Int?,
    hour: Int?, minute: Int, meridiem: String?,
): LocalDateTime {
    var h = hour
    var m = minute
    when {
        h == null -> { h = DEFAULT_HOUR; m = 0 }
        meridiem == "pm" && h < 12 -> h += 12
        meridiem == "am" && h == 12 -> h = 0
        meridiem == null && h in 1..AFTERNOON_UNTIL -> h += 12
    }

    val start = now.withHour(h!!).withMinute(m).withSecond(0).withNano(0)

    if (weekday != null) {
        var ahead = (weekday - start.dayOfWeek.value + 7) % 7
        if (ahead == 0 && !start.isAfter(now)) ahead = 7
        return start.plusDays(ahead.toLong())
    }
    if (dayOffset != null) return start.plusDays(dayOffset.toLong())
    return if (!start.isAfter(now)) start.plusDays(1) else start
}
